client_plugin/drivers/vfile/vfile_driver.go

```diff
@@ -494,6 +494,112 @@ func (d *VolumeDriver) processMount(r volume.MountRequest) volume.Response {
+	var writeEntries []kvstore.KvPair
+	var volRecord etcdops.VFileVolConnectivityData
+	keys := []string{
+		kvstore.VolPrefixInfo + volName,
+	}
+	entries, err := d.kvStore.ReadMetaData(keys)
+	if err != nil {
+		// Failed to fetch existing metadata on the volume
+		// Set volume state to error as we cannot
+		// proceed
+		log.Warningf("Failed to read volume metadata before adding VM to ClientList: %v",
+			err)
+		return err
+	}
+	err = json.Unmarshal([]byte(entries[0].Value), &volRecord)
+	if err != nil {
+		// Failed to unmarshal record from JSON
+		log.Warningf("Failed to unmarshal JSON for reading existing metadata: %v",
+			err)
+		return err
+	}
+	// Add vmIP to CLientList
+	// then marshal the data structure to JSON again.
+	volRecord.ClientList = append(volRecord.ClientList, vmIP)
+	byteRecord, err := json.Marshal(volRecord)
+	if err != nil {
+		// Failed to marshal record as JSON
+		log.Warningf("Failed to marshal JSON for writing metadata: %v",
+			err)
+		return err
+	}
+	writeEntries = append(writeEntries, kvstore.KvPair{
+		Key:   kvstore.VolPrefixInfo + volName,
+		Value: string(byteRecord)})
+
+	log.Infof("Adding VM[%s] to CLientList for %s", vmIP, volName)
+	err = d.kvStore.WriteMetaData(writeEntries)
+	if err != nil {
+		// Failed to write metadata.
+		log.Warningf("Failed to write metadata for volume %s",
+			volName)
+		return err
+	}
+	return nil
+}
+
+func (d *VolumeDriver) removeVMFromClientList(volName string, vmIP string) error {
+	log.Infof("Remove VM[%s] from clientList for volume[%s]", volName, vmIP)
+	var entries []kvstore.KvPair
+	var writeEntries []kvstore.KvPair
+	var volRecord etcdops.VFileVolConnectivityData
+
+	keys := []string{
+		kvstore.VolPrefixInfo + volName,
+	}
+	entries, err := d.kvStore.ReadMetaData(keys)
+	if err != nil {
+		// Failed to fetch existing metadata on the volume
+		// Set volume state to error as we cannot
+		// proceed
+		log.Warningf("Failed to read volume metadata before adding VM to ClientList: %v",
+			err)
+		return err
+	}
+	err = json.Unmarshal([]byte(entries[0].Value), &volRecord)
+	if err != nil {
+		// Failed to unmarshal record from JSON
+		log.Warningf("Failed to unmarshal JSON for reading existing metadata: %v",
+			err)
+		return err
+	}
+	// Add vmIP to CLientList
+	// then marshal the data structure to JSON again.
+	var newClientList []string
+	for _, clientIP := range volRecord.ClientList {
+		if clientIP != vmIP {
+			newClientList = append(newClientList, clientIP)
+		}
+	}
+
+	volRecord.ClientList = newClientList
+	byteRecord, err := json.Marshal(volRecord)
+	if err != nil {
+		// Failed to marshal record as JSON
+		log.Warningf("Failed to marshal JSON for writing metadata: %v",
+			err)
+		return err
+	}
+	writeEntries = append(writeEntries, kvstore.KvPair{
+		Key:   kvstore.VolPrefixInfo + volName,
+		Value: string(byteRecord)})
+
+	log.Infof("Adding VM[%s] to CLientList for %s", vmIP, volName)
+	err = d.kvStore.WriteMetaData(writeEntries)
```

Comment:
Same comment here related to concurrenty. One execution of this method could be overwritten by another execution of this method or addVMToClientList method. For example, invocation-1 is removing an entry and invocation-2 is adding another entry. Does the kvstore support optimistic concurrency control?